src/intellisense/texDocSymbolProvider.ts

```diff
@@ -0,0 +1,296 @@
+import * as vscode from 'vscode';
+import type * as Ast from '@unified-latex/unified-latex-types';
+import * as unifiedLaTeXParse from '@unified-latex/unified-latex-util-parse';
+
+const defaultStructure = ["book", "part", "chapter", "section", "subsection", "subsubsection", "paragraph", "subparagraph"];
+
+const defaultCMD = ["label"];
+
+enum TeXElementType { Environment, Command, Section, SectionAst, SubFile, BibItem, BibField };
+
+type TeXElement = {
+    readonly type: TeXElementType,
+    readonly name: string,
+    label: string,
+    readonly lineFr: number,
+    lineTo: number,
+    readonly filePath: string,
+    children: TeXElement[],
+    parent?: TeXElement,
+    appendix?: boolean
+};
+
+type UnifiedParser = { parse: (content: string) => Ast.Root };
+
+type FileStructureCache = {
+    [filePath: string]: TeXElement[]
+};
+
+let unifiedParser: UnifiedParser = unifiedLaTeXParse.getParser({ flags: { autodetectExpl3AndAtLetter: true } });
+
+export class DocSymbolProvider implements vscode.DocumentSymbolProvider {
+
+    async provideDocumentSymbols(document: vscode.TextDocument): Promise<vscode.DocumentSymbol[]> {
+        const sections = await construct(document);
+        return this.sectionToSymbols(sections);
+    }
+
+    private sectionToKind(section: TeXElement): vscode.SymbolKind {
+        if (section.type === TeXElementType.Section || section.type === TeXElementType.SectionAst) {
+            return vscode.SymbolKind.Struct;
+        }
+        if (section.type === TeXElementType.Environment) {
+            return vscode.SymbolKind.Package;
+        }
+        if (section.type === TeXElementType.Command) {
+            return vscode.SymbolKind.Number;
+        }
+        if (section.type === TeXElementType.SubFile) {
+            return vscode.SymbolKind.File;
+        }
+        if (section.type === TeXElementType.BibItem) {
+            return vscode.SymbolKind.Class;
+        }
+        if (section.type === TeXElementType.BibField) {
+            return vscode.SymbolKind.Constant;
+        }
+        return vscode.SymbolKind.String;
+    }
+
+    private sectionToSymbols(sections: TeXElement[]): vscode.DocumentSymbol[] {
+        const symbols: vscode.DocumentSymbol[] = [];
+
+        sections.forEach(section => {
+            const range = new vscode.Range(section.lineFr, 0, section.lineTo, 65535);
+            const symbol = new vscode.DocumentSymbol(
+                section.label || 'empty', '',
+                this.sectionToKind(section),
+                range, range);
+            symbols.push(symbol);
+            if (section.children.length > 0) {
+                symbol.children = this.sectionToSymbols(section.children);
+            }
+        });
+        return symbols;
+    }
+
+    get triggers(){
+        const latexSelector = ['latex', 'latex-expl3', 'pweave', 'jlweave', 'rsweave'].map( (id) => {
```

Comment:
You are already *listing* the contants and only use them once, why bother `map`?